geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/v1/operations/AbstractRegionRequestOperationHandler.java

```diff
@@ -0,0 +1,76 @@
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.geode.internal.protocol.protobuf.v1.operations;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import org.apache.geode.cache.Region;
+import org.apache.geode.internal.exception.InvalidExecutionContextException;
+import org.apache.geode.internal.protocol.operations.ProtobufOperationHandler;
+import org.apache.geode.internal.protocol.protobuf.v1.BasicTypes;
+import org.apache.geode.internal.protocol.protobuf.v1.Failure;
+import org.apache.geode.internal.protocol.protobuf.v1.MessageExecutionContext;
+import org.apache.geode.internal.protocol.protobuf.v1.ProtobufSerializationService;
+import org.apache.geode.internal.protocol.protobuf.v1.RegionAPI;
+import org.apache.geode.internal.protocol.protobuf.v1.Result;
+import org.apache.geode.internal.protocol.protobuf.v1.Success;
+import org.apache.geode.internal.protocol.protobuf.v1.serialization.exception.DecodingException;
+import org.apache.geode.internal.protocol.protobuf.v1.serialization.exception.EncodingException;
+import org.apache.geode.internal.protocol.protobuf.v1.state.exception.ConnectionStateException;
+
+/**
+ * An abstract base class for requests that operate on a Region. It contains the common logic for
+ * getting a region from a region name, and the operation logic is implemented by child classes.
+ */
+public abstract class AbstractRegionRequestOperationHandler<RequestType, ResponseType>
+    implements ProtobufOperationHandler<RequestType, ResponseType> {
+  private static final Logger logger = LogManager.getLogger();
+
+  @Override
+  public final Result<ResponseType> process(ProtobufSerializationService serializationService,
+      RequestType request, MessageExecutionContext messageExecutionContext)
+      throws InvalidExecutionContextException, EncodingException, DecodingException {
+
+    final String regionName = getRegionName(request);
+
+    Region<Object, Object> region;
+
+    try {
+      region = messageExecutionContext.getCache().getRegion(regionName);
+    } catch (IllegalArgumentException ex) {
+      return Failure.of(BasicTypes.ErrorCode.INVALID_REQUEST,
+          "Invalid region name: \"" + regionName + "\"");
+    }
+
+    if (region == null) {
+      logger.error("Received PutIfAbsentRequest for nonexistent region: {}", regionName);
+      return Failure.of(BasicTypes.ErrorCode.SERVER_ERROR,
+          "Region \"" + regionName + "\" not found");
+    }
+
+    try {
+      messageExecutionContext.getCache().setReadSerializedForCurrentThread(true);
```

Comment:
This is the part I'm least sure about. And actually, we should be restoring state to the value it was before, not to `false` unconditionally.

Reply:
>It would be okay to get the old value and restore it at the end but this thread-local should always be false (absent) in a new thread and should only be set to true for the duration of a method needing to read pdx instances